File: Binary_search.py
def binary_search(lst, item):
    # define initial high and low
    low = 0
    high = len(lst) - 1
    ### lemme see steps for fun/stress testing
    step = 0

    # loop while low and high aren't the same element
    while low <= high:
        ####
        step += 1
        ####
        # define a middle and guess by accessing list
        mid = (low + high) // 2
        guess = lst[mid]
        # did we guess correctly?
        if guess == item:
            print(step)
            return mid

        # are we too low?
        if guess < item:
            # then new low must be at least one above our guess, numbered by mid
            low = mid + 1
        # or are we too high?
        else:
            # then new high must be at least one below our guess
            high = mid - 1
    # if we came down to one number and it wasn't correct, item must not be in list
    return None

# Testing
sorted_list = list(range(75))

binary_search(sorted_list, 6)
binary_search(sorted_list, 12)
binary_search(sorted_list, 11)

File: Binary_search.js
function binary_search(arr, item) {
    let low = 0;
    let high = arr.length - 1;

    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        let guess = arr[mid];

        if (guess == item) return mid;

        guess < item ? low = mid + 1 : high = mid - 1;
    }
    return null;
}

// Testing
const sortedList = [...Array(15).keys()];

console.log(binary_search(sortedList, 4));
console.log(binary_search(sortedList, 9));
console.log(binary_search(sortedList, 11));
